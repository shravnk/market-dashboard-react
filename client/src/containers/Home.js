import React, { Component } from 'react';
import CurrentTable from '../components/CurrentTable'
import PriceHistory from '../components/PriceHistory'
import * as actions from '../actions/stockActions'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

export class Home extends Component  {
  constructor () {
    super()
    this.state = {
      displayHistory: true
    }
  }
  componentWillMount() {
    this.props.actions.clearHistoryData()
  }
  componentDidMount() {
    const stockList = this.props.user.stocks.map(stock => stock.symbol).join(",")
    this.props.actions.fetchStocksCurrentData(stockList)
  }

  changeDisplay = (bool) => {
    this.setState({
      displayHistory: bool
    })
  }

  render() {
    let stocksData = []
    let sectorList = {}
    if (this.props.stocks.currentData) {
      stocksData = this.props.stocks.currentData
      const sectorArr = [...new Set(stocksData.map(s => s.sector))]
      sectorList = arrayToObject(sectorArr)
    }
    let historyData = []
    if (this.props.stocks.historyData) {
      historyData = this.props.stocks.historyData
    }

    return (
      <div className="container-fluid" >
      <div id="Home" className="row">
        <div className="col-md-8" >
          <CurrentTable stocks={stocksData} sectors={sectorList} fetchHistory={this.props.actions.fetchStockHistory}  displayHistory={this.changeDisplay}
          fetchData={this.props.actions.fetchStockDetailedData}/>
        </div>
        <div className="col-md-4 offset-md-8 text-center" style={{position:'fixed'}}>
          <br/>
          <PriceHistory historyData={historyData} symbol={this.props.stocks.historySymbol} display={this.state.displayHistory}
          ready={this.props.stocks.historyReady}/>
        </div>
      </div>
      </div>
    )
  }
}

function mapStateToProps(state) {
  return ({
    stocks: state.stocks,
    user: state.currentUser
  })
}

function mapDispatchToProps(dispatch) {
  return ({
    actions: bindActionCreators(actions, dispatch)
  })
}

const arrayToObject = (array) =>
   array.reduce((obj, item) => {
     obj[item] = item
     return obj
   }, {})


export default connect(mapStateToProps, mapDispatchToProps)(Home)
