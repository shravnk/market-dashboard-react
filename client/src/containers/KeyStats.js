import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from '../actions/stockActions'
import KeyStatsTable from '../components/KeyStatsTable'
import StockHistory from '../components/StockHistory'
import SelectDateRange from '../components/SelectDateRange'


class KeyStats extends Component {

  componentDidMount() {
    this.props.actions.fetchStockDetailedData(this.props.symbol)
    this.props.actions.fetchStockHistory(this.props.symbol, '1D')
  }

  render() {
    let stockData = {}
    if (this.props.stocks.detailedData) {
      stockData = this.props.stocks.detailedData
      return (
        <div className="container">
          <h3 className="text-left">{stockData.companyName} ({stockData.symbol})</h3>
          <SelectDateRange symbol={this.props.stocks.detailedData.symbol} setRange={this.props.actions.fetchStockHistory}/>
          <div className="row" >
            <div className="col-md-5" >
              <KeyStatsTable stockData={stockData} display={true}/>
            </div>
            <div className="col-md-7" >
              <StockHistory historyData={this.props.stocks.historyData} period={this.props.stocks.historyPeriod}/>
            </div>
          </div>
        </div>
      )
    } else {
      return (
        <p>Loading...</p>
      )
    }
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    symbol: ownProps.match.params.symbol,
    stocks: state.stocks
  }
}

function mapDispatchToProps(dispatch) {
  return ({
    actions: bindActionCreators(actions, dispatch)
  })
}

export default connect(mapStateToProps, mapDispatchToProps)(KeyStats)
