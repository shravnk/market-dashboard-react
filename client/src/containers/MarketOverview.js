import React, { Component } from 'react';
import Gainers from '../components/Gainers'
import Losers from '../components/Losers'
import IndexCurrent from '../components/IndexCurrent'
import IndexHistory from '../components/IndexHistory'
import * as actions from '../actions/stockActions'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { withRouter } from 'react-router-dom'



class MarketOverview extends Component  {

  componentDidMount() {
    this.props.actions.fetchGainersData()
    this.props.actions.fetchLosersData()
    this.props.actions.fetchIndicesCurrentData()
    this.props.actions.fetchIndexHistory()
  }

  render() {
    const loading = this.props.indexHistory === undefined || this.props.indexHistory.length == 0
    if (loading) {
      return (
        <p>Loading...</p>
      )
    } else {
    return (
      <div className="container-fluid" >
        <div id="indices" className="row" >
          <div className="col-md-4" >
            <IndexCurrent indices={this.props.indices} changeIndexDisplay={this.props.actions.changeIndexDisplay}/>
          </div>
          <div className="col-md-8" >
            <IndexHistory indexData={this.props.indexHistory[this.props.currentIndex]} />
          </div>
        </div>
        <div id="Movers" className="row">
          <div className="col-md-6" >
            <h5>Best Performers</h5>
            <Gainers gainers={this.props.gainers} />
          </div>
          <div className="col-md-6" >
            <h5>Worst Performers</h5>
            <Losers losers={this.props.losers} />
          </div>
        </div>
      </div>
    )
    }
  }
}

function mapStateToProps(state) {
  return ({
    gainers: state.stocks.gainerData,
    losers: state.stocks.loserData,
    indices: state.stocks.indexData,
    indexHistory: state.stocks.indexHistory,
    currentIndex: state.stocks.currentIndexName
  })
}

function mapDispatchToProps(dispatch) {
  return ({
    actions: bindActionCreators(actions, dispatch)
  })
}


export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MarketOverview))
