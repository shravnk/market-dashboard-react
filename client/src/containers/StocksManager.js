import React, {Component} from 'react';
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { bindActionCreators } from 'redux'
import AddStocksForm from '../components/AddStocksForm'
import {addStocks} from '../actions/userActions.js'


class StocksManager extends Component {
  render() {
    return(
      <AddStocksForm currentUser={this.props.currentUser} addStocks={this.props.addStocks} />
    )
  }
}

function mapStateToProps(state) {
  return ({
    currentUser: state.currentUser.username
  })
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    addStocks: addStocks}
    , dispatch)
}


export default withRouter(connect(mapStateToProps, mapDispatchToProps)(StocksManager))
