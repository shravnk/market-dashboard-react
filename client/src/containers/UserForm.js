import React, {Component} from 'react';
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import UserLogin from '../components/UserLogin'
import UserSignup from '../components/UserSignup'
import {newUser} from '../actions/userActions.js'


class UserForm extends Component {
  render() {

    const renderForm = () => {
      if ( this.props.match.path === "/login"){
          return (<UserLogin loginUser={this.props.loginUser}/>)
        } else if ( this.props.match.path === "/signup") {
          return(<UserSignup createUser={this.props.newUser} />)
      }
    }
    return(
      renderForm()
    )
  }
}
const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    newUser: newUser,
    loginUser: loginUser}
    , dispatch)
}

export default connect(null, mapDispatchToProps)(UserForm)
