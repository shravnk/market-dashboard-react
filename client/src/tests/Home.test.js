import React from 'react'
import { Home } from '../../src/containers/Home'

const mockFn = jest.fn()
const mockFetchCurrentData = jest.fn()
const mockClearHistoryData = jest.fn()

const mockSymbol = "T"
const mockHistory = ["A", "B", "C"]

const props = {
                  actions: {
                    clearHistoryData: mockClearHistoryData,
                    fetchStocksCurrentData: mockFetchCurrentData,
                    fetchStockHistory: mockFn,
                    fetchStockDetailedData: mockFn
                  },
                  stocks: {
                    historyData: mockHistory,
                    currentData: [],
                    historySymbol: mockSymbol,
                    historyReady: false
                  },
                  user: {
                    stocks: []
                  }
              }

const wrapper = shallow(<Home {...props}/>)

test('component renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
});

test('state.displayHistory set to true', () => {
    expect(wrapper.state('displayHistory')).toEqual(true)
});

test('only clearHistoryData and fetchStocksCurrentData are called after mounting', () => {
    expect(mockFn).not.toHaveBeenCalled()
    expect(mockClearHistoryData).toHaveBeenCalled()
    expect(mockFetchCurrentData).toHaveBeenCalled()
})

test('passes correct props to CurrentTable Component', () => {
  const expectedProps = {stocks: [], sectors: {}, fetchHistory: mockFn, displayHistory: expect.any(Function), fetchData: mockFn}
  const currentTableProps = wrapper.find('withRouter(CurrentTable)').props()
  expect(currentTableProps).toEqual(expectedProps)
})

test('passes correct props to PriceHistory Component', () => {
  const expectedProps = { symbol: mockSymbol, historyData: mockHistory, display: true, ready: false }
  const priceHistoryProps = wrapper.find('PriceHistory').props()
  expect(priceHistoryProps).toEqual(expectedProps)
})
