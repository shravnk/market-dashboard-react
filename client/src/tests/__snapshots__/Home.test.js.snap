// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    actions={
        Object {
            "clearHistoryData": [Function],
            "fetchStockDetailedData": [Function],
            "fetchStockHistory": [Function],
            "fetchStocksCurrentData": [Function],
          }
    }
    stocks={
        Object {
            "currentData": Array [],
            "historyData": Array [
              "A",
              "B",
              "C",
            ],
            "historyReady": false,
            "historySymbol": "T",
          }
    }
    user={
        Object {
            "stocks": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="row"
        id="Home"
>
        <div
                className="col-md-8"
        >
                <withRouter(CurrentTable)
                        displayHistory={[Function]}
                        fetchData={[Function]}
                        fetchHistory={[Function]}
                        sectors={Object {}}
                        stocks={Array []}
                />
        </div>
        <div
                className="col-md-4 offset-md-8 text-center"
                style={
                        Object {
                                "position": "fixed",
                              }
                }
        >
                <br />
                <PriceHistory
                        display={true}
                        historyData={
                                Array [
                                        "A",
                                        "B",
                                        "C",
                                      ]
                        }
                        ready={false}
                        symbol="T"
                />
        </div>
</div>,
      "className": "container-fluid",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="col-md-8"
>
            <withRouter(CurrentTable)
                        displayHistory={[Function]}
                        fetchData={[Function]}
                        fetchHistory={[Function]}
                        sectors={Object {}}
                        stocks={Array []}
            />
</div>,
          <div
            className="col-md-4 offset-md-8 text-center"
            style={
                        Object {
                                    "position": "fixed",
                                  }
            }
>
            <br />
            <PriceHistory
                        display={true}
                        historyData={
                                    Array [
                                                "A",
                                                "B",
                                                "C",
                                              ]
                        }
                        ready={false}
                        symbol="T"
            />
</div>,
        ],
        "className": "row",
        "id": "Home",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <withRouter(CurrentTable)
              displayHistory={[Function]}
              fetchData={[Function]}
              fetchHistory={[Function]}
              sectors={Object {}}
              stocks={Array []}
/>,
            "className": "col-md-8",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "displayHistory": [Function],
              "fetchData": [Function],
              "fetchHistory": [Function],
              "sectors": Object {},
              "stocks": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <br />,
              <PriceHistory
                display={true}
                historyData={
                                Array [
                                                "A",
                                                "B",
                                                "C",
                                              ]
                }
                ready={false}
                symbol="T"
/>,
            ],
            "className": "col-md-4 offset-md-8 text-center",
            "style": Object {
              "position": "fixed",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "display": true,
                "historyData": Array [
                  "A",
                  "B",
                  "C",
                ],
                "ready": false,
                "symbol": "T",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
          id="Home"
>
          <div
                    className="col-md-8"
          >
                    <withRouter(CurrentTable)
                              displayHistory={[Function]}
                              fetchData={[Function]}
                              fetchHistory={[Function]}
                              sectors={Object {}}
                              stocks={Array []}
                    />
          </div>
          <div
                    className="col-md-4 offset-md-8 text-center"
                    style={
                              Object {
                                        "position": "fixed",
                                      }
                    }
          >
                    <br />
                    <PriceHistory
                              display={true}
                              historyData={
                                        Array [
                                                  "A",
                                                  "B",
                                                  "C",
                                                ]
                              }
                              ready={false}
                              symbol="T"
                    />
          </div>
</div>,
        "className": "container-fluid",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-md-8"
>
              <withRouter(CurrentTable)
                            displayHistory={[Function]}
                            fetchData={[Function]}
                            fetchHistory={[Function]}
                            sectors={Object {}}
                            stocks={Array []}
              />
</div>,
            <div
              className="col-md-4 offset-md-8 text-center"
              style={
                            Object {
                                          "position": "fixed",
                                        }
              }
>
              <br />
              <PriceHistory
                            display={true}
                            historyData={
                                          Array [
                                                        "A",
                                                        "B",
                                                        "C",
                                                      ]
                            }
                            ready={false}
                            symbol="T"
              />
</div>,
          ],
          "className": "row",
          "id": "Home",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <withRouter(CurrentTable)
                displayHistory={[Function]}
                fetchData={[Function]}
                fetchHistory={[Function]}
                sectors={Object {}}
                stocks={Array []}
/>,
              "className": "col-md-8",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "displayHistory": [Function],
                "fetchData": [Function],
                "fetchHistory": [Function],
                "sectors": Object {},
                "stocks": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <br />,
                <PriceHistory
                  display={true}
                  historyData={
                                    Array [
                                                      "A",
                                                      "B",
                                                      "C",
                                                    ]
                  }
                  ready={false}
                  symbol="T"
/>,
              ],
              "className": "col-md-4 offset-md-8 text-center",
              "style": Object {
                "position": "fixed",
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "display": true,
                  "historyData": Array [
                    "A",
                    "B",
                    "C",
                  ],
                  "ready": false,
                  "symbol": "T",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
